lib/edgehog_device_forwarder/connections.ex

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Connections do
+  use Supervisor
+
+  alias EdgehogDeviceForwarder.Connections.TokenData
+
+  @opts Application.compile_env!(:edgehog_device_forwarder, EdgehogDeviceForwarder.Connections)
+  @tokens @opts[:tokens]
+  @messages @opts[:messages]
+
+  @type token :: String.t()
+  @type device_id :: TokenData.device_id()
+  @type realm :: TokenData.realm()
+  @type socket :: TokenData.socket()
+  @type controller :: pid()
+  @type http_request_id :: binary()
+  @type ws_socket_id :: binary()
+
+  # API
+
+  ## Token
+  @spec register_device_info(token, %{:device_id => device_id, :realm => realm}) ::
+          :ok | {:error, :token_not_found}
+  def register_device_info(token, device_info) do
+    ConCache.update(@tokens, token, &TokenData.put_device_info(&1, device_info))
+  end
+
+  @spec register_socket(token, socket) :: :ok | {:error, :already_exists}
+  def register_socket(token, socket) do
+    # TODO: check if we need to handle socket's process termination ourself.
+    # I'd be inclined to say no from
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#c:terminate/2
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#module-custom-transports
+    ConCache.insert_new(@tokens, token, TokenData.new(socket))
+  end
+
+  @spec close_socket(token) :: :ok
+  def close_socket(token) do
+    ConCache.delete(@tokens, token)
+  end
+
+  @spec fetch_socket(token) :: {:ok, socket} | {:error, :token_not_found}
+  def fetch_socket(token) do
+    case ConCache.get(@tokens, token) do
+      nil -> {:error, :token_not_found}
+      %TokenData{socket: socket} -> {:ok, socket}
+    end
+  end
+
+  ## Messages
+
+  @spec new_http_request(controller) :: {:ok, http_request_id}
+  def new_http_request(controller) do
```

Comment:
The controller doesn't have a `pid`, it's the request that lives inside a process, so I think you have to rework naming here

Reply:
In #11 I actually went with [saving the controller's pid](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/11/commits/a0620cd7391f26c6d3ae24f69447daaf167d2dcc#diff-c2f32d20e53f7460e7e01b39b7003c2c8837e763daf382304d668254e4c438a9R30) and [waiting for a response](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/11/commits/a0620cd7391f26c6d3ae24f69447daaf167d2dcc#diff-c2f32d20e53f7460e7e01b39b7003c2c8837e763daf382304d668254e4c438a9R38).

This way we can look at the forwarder as just another function call and leave all the `Plug.Conn` logic inside the controller